### DISCUSSION TITLE
Can I register a `ValueConverter` for a struct globally?

### DISCUSSION THREAD

**Discussion by timbze** [linq2db/linq2db#3814] [Q&A] [answered]
I have a struct that I want to use for a POCO property type, but the only way I can get it working is this:
```csharp
[Column]
[ValueConverter(ConverterType = typeof(DataId.LinqToDbValueConverter))]
public DataId StringId { get; set; } // DataId is the struct
```

I would like to register the ValueConverter globally, but I have not found any way to make it work. The ValueConverter is this:
```csharp
public class LinqToDbValueConverter : LinqToDB.Common.ValueConverter<DataId, string>
{
    public LinqToDbValueConverter()
        : base(
              v => v.Value,
              p => DataId.Deserialize(p),
              handlesNulls: false)
    { }
}
```

This, which I _figured_ should work, is NOT working:
```csharp
// register type globally
Converter.SetConverter<DataId, string>(d => d.Value);
Converter.SetConverter<string, DataId>(s => DataId.From(s));
```

The error I get when I have set `Converter.SetConverter` is this:
```
System.NotSupportedException: The CLR type DataId isn't natively supported by Npgsql or your PostgreSQL. To use it with a PostgreSQL composite you need to specify DataTypeName or to map it, please refer to the documentation.
```

Is there any way I can register a `DataId` converter globally so that I don't have to add a `ValueConverter` attribute to every property that uses the type?

**Comment by viceroypenguin** [comment]
You'll want to create a static `MappingSchema` that is used for all instances of your custom `DataContext`. When creating the `MappingSchema`, you can add a converter using `schema.SetConverter<DataId, string>(d => d.Value);` and vice-versa.

**Reply by timbze** [reply]
I found out now that it works for reading data, but not for `Insert()` extensions. I guess it's a bug.

**Comment by timbze** [comment, marked as answer]
Created #3815. I think there's a bug with the way it seems that it should work

**EDIT**
Was not a bug. I needed a 3rd converter.
`Converter.SetConverter<DataId, DataParameter>(d => new DataParameter { DataType = DataType.Text, Value = d.Value, });`
